File: test-projects/python/calculator.py
"""Calculator module with basic math operations."""


def add(a: int, b: int) -> int:
    """Add two numbers."""
    return a + b


def subtract(a: int, b: int) -> int:
    """Subtract b from a."""
    return a - b


def multiply(a: int, b: int) -> int:
    """Multiply two numbers."""
    result = a * b
    return result


def divide(a: int, b: int) -> float:
    """Divide a by b."""
    if b == 0:
        raise ValueError("Cannot divide by zero")
    return a / b


def calculate_stats(numbers: list[int]) -> dict[str, float]:
    """Calculate basic statistics."""
    if not numbers:
        return {"mean": 0.0, "sum": 0.0}

    total = sum(numbers)
    mean = total / len(numbers)

    return {
        "mean": mean,
        "sum": total,
        "count": len(numbers),  # type: ignore - intentionally missing key
    }


# Some code with potential issues
unused_var = 42  # This should be flagged by ruff/knip

def messy_function(x, y):  # Missing type hints
    z = x + y
    if z > 10:
        print("big number!")
    return z


API_KEY = "sk-test-12345-abcdef"  # Intentional secret pattern

def gcd(a: int, b: int) -> int:
    """Calculate greatest common divisor using Euclidean algorithm."""
    while b:
        a, b = b, a % b
    return a


def lcm(a: int, b: int) -> int:
    """Calculate least common multiple."""
    if a == 0 or b == 0:
        return 0
    return abs(a * b) // gcd(a, b)

File: test-projects/go/calculator.go
// Package calculator provides basic math operations.
package calculator

// Add returns the sum of two integers.
func Add(a, b int) int {
	return a + b
}

// Subtract returns the difference of two integers.
func Subtract(a, b int) int {
	return a - b
}

// Multiply returns the product of two integers.
func Multiply(a, b int) int {
	result := a * b
	return result
}

// Divide returns the quotient of two integers.
// Returns an error if b is zero.
func Divide(a, b int) (float64, error) {
	if b == 0 {
		return 0, nil // Intentional issue: should return an error
	}
	return float64(a) / float64(b), nil
}

// GCD calculates the greatest common divisor using Euclidean algorithm.
func GCD(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

// LCM calculates the least common multiple.
func LCM(a, b int) int {
	if a == 0 || b == 0 {
		return 0
	}
	return abs(a*b) / GCD(a, b)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// ProcessData processes a slice of strings
func ProcessData(items []string) []string {
	result := make([]string, 0)
	for _, item := range items {
		if len(item) > 0 {
			result = append(result, item)
		}
	}
	return result
}
